deps/aports/download-repos.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
+import tqdm
+import urllib.request
+import urllib.parse
+import tarfile
+import pathlib
+import concurrent.futures
+
+IX_NAME = 'P'
+IX_VERSION = 'V'
+
+def read_index(index):
+    pkgs = []
+    pkg = {}
+    while True:
+        line = index.readline()
+        oline = line
+        if line == b'\n':
+            pkgs.append(pkg)
+            pkg = {}
+            continue
+        if line == b'':
+            break
+        assert line.endswith(b'\n')
+        line = line[:-1]
+
+        key, _, value = line.partition(b':')
+        pkg[key.decode()] = value.decode()
+    return pkgs
+
+def download_repo(root_url, repo_name, index_name):
+    repo = pathlib.Path(repo_name)
+    repo.mkdir(parents=True, exist_ok=True)
+    index_path = repo/'APKINDEX.tar.gz'
+    index_url = f'{root_url}/{repo_name}/{index_name}'
+    urllib.request.urlretrieve(index_url, index_path)
+    with tarfile.open(index_path) as tar:
+        pkgs = read_index(tar.extractfile('APKINDEX'))
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=30) as pool:
+        futures = [pool.submit(download_package, repo, f'{root_url}/{repo_name}', pkg) for pkg in pkgs]
+        for f in tqdm.tqdm(concurrent.futures.as_completed(futures), total=len(futures)):
+            f.result()
+
+def download_package(repo, root_url, pkg):
```

Comment:
throttling is a thing you should do

Reply:
Why

Reply:
It's polite to the servers, since it seems like you're firing off thousands of requests.

Reply:
> Backblaze recommends that you do lots of transfers simultaneously for maximum speed.
https://rclone.org/b2/

Reply:
"please DDOS me"…well, if that's what they want

Reply:
It's only 30 requests at once